CIP-????/README.md

```diff
@@ -0,0 +1,709 @@
+
+A more complex design would be to have the bindings for the variables be part of the case expression instead. That is, the current design does this:
+
+```
+(case (constr 1 a b) (\x -> t1) (\x y -> t2))
+```
+(the literal lambdas here could be arbitrary terms)
+
+whereas we could do this:
+
+```
+(case (constr 1 x y) (alt x t1) (alt x y t2))
+```
+
+The effect of this is that we know statically that we have a function, and we can jump right into evaluating the _body_ of the case branch instead of first having to evaluate it to a function and apply the arguments.
+
+This saves us a small but meaningful amount of overhead at the cost of making the implementation significantly more complex.
+
+We prototyped this version and it was noticeably faster (~10%).
+However, performance investigations showed that we can realize a significant amount of the performance gain through other means, leaving only about 3% unclaimed.[^see-appendix-2]
+We think that this is an acceptable cost for a simpler implementation.
+
+[^see-appendix-2]: See Appendix 2 for more details.
+
+#### 3. Convert Redeemers and Datums to Sums-of-Products Rather Than Builtin Data
+
+This proposal changes how the script context is passed to scripts, thus removing one use of the Plutus builtin type `Data`.
+However, we still use the builtin `Data` type for passing redeemers and datums, which continue to be `Data` objects in the transaction.[^why-redeemers-datums-data]
+
+[^why-redeemers-datums-data]: See "Why not also make datums and redeemers terms instead of `Data`?" for why we want to keep redeemers and datums as `Data` in transactions themselves.
+
+An alternative would be to encode the `Data` objects for redeemers and datums using sums-of-products instead of the builtin `Data` type.
+That would require us to define and specify a stable `dataToTerm :: Data -> Term` function that converts a `Data` object into a term using sums-of-products.
+This would be simple for some cases (e.g. `Constr`), but less simple for other cases (notably `Map` and `List`).
+
+This has some costs:
+1. We have an additional conversion function that we need to specify.
+2. The details of the conversion function would be relevant to script authors, since they would need to know how their `Data` object had been encoded in order to operate on it.
+3. The ledger-script interface becomes generally more complex, since instead of passing `Data` objects directly, we now need to encode them.
+
+However, it would have a significant advantage in that we would potentially be able to remove the builtin `Data` type and its associated builtins entirely.
+Of course, they would still have to be supported in old ledger languages, so the reduction in implementation complexity would be limited.
+
+Strictly, this could be implemented as a later change, but if we're going to change the ledger-script interface anyway, it would be nice to do it in one go.
```

Comment:
Yes please! 🙏

Reply:
> Loosing the homogeneity of validator's arguments is a massive loss.

Why? I don't see how this is a big deal. I think it's mostly aesthetically displeasing but I can't think of any practical issues?

Reply:
It has practical issues for us with Aiken (and I reckon other languages with a similar approach like Helios). Because we do not decode the script context, datums or redeemers. Aiken types are encoded as `Data` directly. So when one pulls a value of type `A` from the datum, and matches it with a value of type `A` in the script context, there's no conversion needed because both arguments have the same structural format. 

Now, if the script context is no longer a `Data` builtin type but a completely new term. That means we must either encode the script context as `Data` (likely no the preferred way because it can be quite large) or convert the datum / redeemer into terms (getting rid of the `Data` wrapper). 

Besides I really don't see any benefits in preserving `Data` on-chain, in the script evaluation. With the introduction of `constr`, any `Data` can be expressed as simpler terms, which means that an on-chain validator would be better off (performance wise) manipulating those less expensive terms directly instead of dealing with `Data`. 

`Data` is only truly useful on-chain if we need to have non-homogeneous data structures (like a array of _things_) like one might have in dynamically typed languages. I am not sure if I have any use case for it though to be frank but my brain is perhaps too used to strong static typing now. 

There's possibly a "security" argument in **not** translating a `Data` as an SOP. Because this translation would have to be done by the ledger, it means that there's no direct cost for it. When it's done as part of the script execution, then the cost of this translation is paid through the script execution. However, I'd argue that this cost is _implicitly_ covered by the standard fees. The larger the datum / redeemer, the larger the transaction and the higher the fees. So datums that are more computationally expensive to translate also comes with larger transaction fees. Thus, it doesn't introduce any new attack vectors in that sense. Plus, the cost of doing this translation is Haskell can probably be quantified and is in all likelihood negligible in front of e.g. a digital signature verification.

Reply:
> Aiken types are encoded as Data directly

I agree it's a problem for Aiken if we stop encoding things in the way that Aiken expects. I don't really know what to say to people who have come to rely on (what I perceive as) incidental features of the status quo. I don't see an easy way to improve things at all without breaking some assumptions people will have about the way things work :| 

I also don't see how consistency helps you then, surely that just makes your problem worse?

Do you see any non-Aiken problems?

> Data is only truly useful on-chain if we need to have non-homogeneous data structures (like a array of things) like one might have in dynamically typed languages.

See "Why not also make datums and redeemers terms instead of `Data`?" for why redeemers and datums are still Data. Given that, the simplest thing to do is to represent them exactly as they are.

> Plus, the cost of doing this translation is Haskell can probably be quantified and is in all likelihood negligible in front of e.g. a digital signature verification.

Haha, we made that argument in the past and it turned out to be false. We have an extremely limited budget for anything script related, we _have to_ measure it.

I do think the argument that it would be incentive-compatible is reasonable though: users already pay a fee proportional to the size of the datum/redeemer in order to post it. Except for inline datums :thinking:

Reply:
I don't believe this is an Aiken-specific issue. PlutusTx users make use of the same tricks.

Reply:
> I don't believe this is an Aiken-specific issue. PlutusTx users make use of the same tricks.

Such as? What Matthias wrote seemed quite Aiken-specific to me, since it's about how Aiken encodes datatypes?

(Incidentally I benchmarked this, and it seems to indeed be much faster to encode datatypes as SOPs (or even Scott-encoded) than using `Data`. See the text for more.)

Reply:
Using toBuiltinData allows you to make equalsData comparisons. Which I’ve used in PlutusTx to circumvent slower comparisons when comparing txOutRef or Value.